diff --git a/locale/loadarchive.c b/locale/loadarchive.c
index ba0fe456..38f82117 100644
--- a/locale/loadarchive.c
+++ b/locale/loadarchive.c
@@ -43,6 +43,7 @@
 
 /* Name of the locale archive file.  */
 static const char archfname[] = COMPLOCALEDIR "/locale-archive";
+static const char nixDefaultLocaleArchive[] = "/nix/lib/locale/locale-archive";
 
 /* Size of initial mapping window, optimal if large enough to
    cover the header plus the initial locale.  */
@@ -123,6 +124,25 @@ calculate_head_size (const struct locarhead *h)
   return MAX (namehash_end, MAX (string_end, locrectab_end));
 }
 
+static int
+open_locale_archive (void)
+{
+  int fd = -1;
+  char *versioned_path = getenv ("LOCALE_ARCHIVE_2_27");
+  char *path = getenv ("LOCALE_ARCHIVE");
+  if (versioned_path)
+    fd = __open_nocancel (versioned_path, O_RDONLY|O_LARGEFILE|O_CLOEXEC);
+  if (path && fd < 0)
+    fd = __open_nocancel (path, O_RDONLY|O_LARGEFILE|O_CLOEXEC);
+  if (fd < 0)
+    fd = __open_nocancel (archfname, O_RDONLY|O_LARGEFILE|O_CLOEXEC);
+  if (fd < 0)
+    fd = __open_nocancel (nixDefaultLocaleArchive, O_RDONLY|O_LARGEFILE|O_CLOEXEC);
+  if (fd < 0)
+    fd = __open_nocancel ("/usr/lib/locale/locale-archive", O_RDONLY|O_LARGEFILE|O_CLOEXEC);
+  return fd;
+}
+
 
 /* Find the locale *NAMEP in the locale archive, and return the
    internalized data structure for its CATEGORY data.  If this locale has
@@ -202,7 +222,7 @@ _nl_load_locale_from_archive (int category, const char **namep)
       archmapped = &headmap;
 
       /* The archive has never been opened.  */
-      fd = __open_nocancel (archfname, O_RDONLY|O_LARGEFILE|O_CLOEXEC);
+      fd = open_locale_archive ();
       if (fd < 0)
 	/* Cannot open the archive, for whatever reason.  */
 	return NULL;
@@ -397,8 +417,7 @@ _nl_load_locale_from_archive (int category, const char **namep)
 	  if (fd == -1)
 	    {
 	      struct stat64 st;
-	      fd = __open_nocancel (archfname,
-				    O_RDONLY|O_LARGEFILE|O_CLOEXEC);
+	      fd = open_locale_archive ();
 	      if (fd == -1)
 		/* Cannot open the archive, for whatever reason.  */
 		return NULL;
diff --git a/locale/programs/locale.c b/locale/programs/locale.c
index e2e309c2..64923735 100644
--- a/locale/programs/locale.c
+++ b/locale/programs/locale.c
@@ -47,6 +47,7 @@
 #include <programs/xmalloc.h>
 
 #define ARCHIVE_NAME COMPLOCALEDIR "/locale-archive"
+#define NIX_DEFAULT_LOCALE_ARCHIVE "/nix/lib/locale/locale-archive"
 
 /* If set print the name of the category.  */
 static int show_category_name;
@@ -633,6 +634,26 @@ nameentcmp (const void *a, const void *b)
 }
 
 
+static int
+open_locale_archive (void)
+{
+  int fd = -1;
+  char *versioned_path = getenv ("LOCALE_ARCHIVE_2_27");
+  char *path = getenv ("LOCALE_ARCHIVE");
+  if (versioned_path)
+    fd = open64 (versioned_path, O_RDONLY);
+  if (path && fd < 0)
+    fd = open64 (path, O_RDONLY);
+  if (fd < 0)
+    fd = open64 (ARCHIVE_NAME, O_RDONLY);
+  if (fd < 0)
+    fd = open64 (NIX_DEFAULT_LOCALE_ARCHIVE, O_RDONLY);
+  if (fd < 0)
+    fd = open64 ("/usr/lib/locale/locale-archive", O_RDONLY);
+  return fd;
+}
+
+
 static int
 write_archive_locales (void **all_datap, char *linebuf)
 {
@@ -645,7 +666,7 @@ write_archive_locales (void **all_datap, char *linebuf)
   int fd, ret = 0;
   uint32_t cnt;
 
-  fd = open64 (ARCHIVE_NAME, O_RDONLY);
+  fd = open_locale_archive ();
   if (fd < 0)
     return 0;
 
diff --git a/locale/programs/locarchive.c b/locale/programs/locarchive.c
index dccaf04e..b52787c9 100644
--- a/locale/programs/locarchive.c
+++ b/locale/programs/locarchive.c
@@ -58,6 +58,7 @@
 extern const char *output_prefix;
 
 #define ARCHIVE_NAME COMPLOCALEDIR "/locale-archive"
+#define NIX_DEFAULT_LOCALE_ARCHIVE "/nix/lib/locale/locale-archive"
 
 static const char *locnames[] =
   {
@@ -117,6 +118,24 @@ prepare_address_space (int fd, size_t total, size_t *reserved, int *xflags,
 }
 
 
+static int
+open_locale_archive (const char * archivefname, int flags)
+{
+  int fd = -1;
+  char *versioned_path = getenv ("LOCALE_ARCHIVE_2_27");
+  char *path = getenv ("LOCALE_ARCHIVE");
+  if (versioned_path)
+    fd = open64 (versioned_path, flags);
+  if (path && fd < 0)
+    fd = open64 (path, flags);
+  if (fd < 0)
+    fd = open64 (NIX_DEFAULT_LOCALE_ARCHIVE, flags);
+  if (fd < 0)
+    fd = open64 (archivefname, flags);
+  return fd;
+}
+
+
 static void
 create_archive (const char *archivefname, struct locarhandle *ah)
 {
@@ -578,7 +597,7 @@ open_archive (struct locarhandle *ah, bool readonly)
   while (1)
     {
       /* Open the archive.  We must have exclusive write access.  */
-      fd = open64 (archivefname, readonly ? O_RDONLY : O_RDWR);
+      fd = open_locale_archive (archivefname, readonly ? O_RDONLY : O_RDWR);
       if (fd == -1)
 	{
 	  /* Maybe the file does not yet exist? If we are opening
diff --git a/sysdeps/generic/unsecvars.h b/sysdeps/generic/unsecvars.h
index 5ea8a4a2..1eb821ad 100644
--- a/sysdeps/generic/unsecvars.h
+++ b/sysdeps/generic/unsecvars.h
@@ -24,6 +24,8 @@
   "LD_SHOW_AUXV\0"							      \
   "LD_USE_LOAD_BIAS\0"							      \
   "LOCALDOMAIN\0"							      \
+  "LOCALE_ARCHIVE\0"							      \
+  "LOCALE_ARCHIVE_2_27\0"						      \
   "LOCPATH\0"								      \
   "MALLOC_TRACE\0"							      \
   "NIS_PATH\0"								      \
